sno/dataset2.py

```diff
@@ -122,6 +125,141 @@ def hexhash(self):
         return _hexhash(self.dumps())
 
 
+def _pk_index_ordering(column):
+    """Returns primary key columns first, in pk_index order, then other columns."""
+    if column.pk_index is not None:
+        return column.pk_index
+    else:
+        # Non primary-key columns after primary key columns.
+        return float('inf')
+
+
+class ColumnSchema:
+    """
+    The schema for a single column. A column has a unique ID that is constant for the columns lifetime
+    - even if the column is moved or renamed - and it has a name, a datatype, info about if this
+    column is one of the primary keys, and potentially extra info about the specific datatype.
+    """
+
+    def __init__(self, id, name, data_type, pk_index=None, extra_type_info=None):
+        self.id = id
+        self.name = name
+        self.data_type = data_type
+        self.pk_index = pk_index
+        self.extra_type_info = extra_type_info if extra_type_info is not None else {}
+
+    @classmethod
+    def from_json_dict(cls, json_dict):
+        return cls(
+            json_dict.pop("id"),
+            json_dict.pop("name"),
+            json_dict.pop("dataType"),
+            json_dict.pop("primaryKeyIndex"),
+            json_dict,
+        )
+
+    def to_json_dict(self):
+        return {
+            "id": self.id,
+            "name": self.name,
+            "dataType": self.data_type,
+            "primaryKeyIndex": self.pk_index,
+            **self.extra_type_info,
+        }
+
+    def __eq__(self, other):
+        if not isinstance(other, ColumnSchema):
+            return False
+        return (
+            self.id == other.id
+            and self.name == other.name
+            and self.data_type == other.data_type
+            and self.pk_index == other.pk_index
+            and self.extra_type_info == other.extra_type_info
+        )
+
+
+class Schema:
+    """A schema is just an ordered list of ColumnSchemas."""
```

Comment:
Schema and Legend classes are immutable, right? Do we want to enforce that? Or alternatively, document it?

Reply:
Enforced, more-or-less.